modules/tinymce/src/themes/silver/main/ts/ui/core/complex/BespokeNumberInput.ts

```diff
@@ -0,0 +1,185 @@
+import { onControlAttached, onControlDetached } from '../../controls/Controls';
+import { updateMenuText, UpdateMenuTextEvent } from '../../dropdown/CommonDropdown';
+import { onSetupEvent } from '../ControlUtils';
+import { NumberInputSpec } from './FontSizeBespoke';
+
+interface BespokeSelectApi {
+  readonly getComponent: () => AlloyComponent;
+}
+
+const createBespokeNumberInput = (editor: Editor, backstage: UiFactoryBackstage, spec: NumberInputSpec): AlloySpec => {
+  let currentComp: Optional<AlloyComponent> = Optional.none();
+
+  const getValueFromCurrentComp = (comp: Optional<AlloyComponent>): string =>
+    comp.map((alloyComp) => Representing.getValue(alloyComp)).getOr('');
+
+  const onSetup = onSetupEvent(editor, 'NodeChange', (api: BespokeSelectApi) => {
+    const comp = api.getComponent();
+    currentComp = Optional.some(comp);
+    spec.updateInputValue(comp);
+  });
+
+  const getApi = (comp: AlloyComponent): BespokeSelectApi => ({ getComponent: Fun.constant(comp) });
+  const editorOffCell = Cell(Fun.noop);
+
+  const customEvents = Id.generate('custom-number-input-events');
+
+  const isValidValue = (value: number): boolean => value >= 0;
+
+  const changeValue = (f: (v: number, step: number) => number, focusBack: boolean): void => {
+    const text = getValueFromCurrentComp(currentComp);
+    const parsedText = Dimension.parse(text, [ 'unsupportedLength' ]);
+    const value = parsedText.map((res) => res.value).getOr(0);
+    const unit = parsedText.map((res) => res.unit).getOr('');
+    const newValue = f(value, spec.getConfigFromUnit(unit).step);
+    const newValueWithUnit = `${isValidValue(newValue) ? newValue : value}${unit}`;
+
+    spec.onAction(newValueWithUnit, focusBack);
+    currentComp.each((comp) => Representing.setValue(comp, newValueWithUnit));
+  };
+
+  const decrease = (focusBack: boolean) => changeValue((n, s) => n - s, focusBack);
+  const increase = (focusBack: boolean) => changeValue((n, s) => n + s, focusBack);
+
+  const goToParent = (comp: AlloyComponent) =>
+    Traverse.parentElement(comp.element).fold(Optional.none, (parent) => {
+      Focus.focus(parent);
+      return Optional.some(true);
+    });
+
+  const memInput = Memento.record({
+    dom: {
+      tag: 'div',
+      classes: [ 'tox-input-wrapper', 'highlight-on-focus' ]
+    },
+    components: [
+      Input.sketch({
+        inputBehaviours: Behaviour.derive([
+          AddEventsBehaviour.config(customEvents, [
+            onControlAttached({ onSetup, getApi }, editorOffCell),
+            onControlDetached({ getApi }, editorOffCell)
+          ]),
+          AddEventsBehaviour.config('menubutton-update-display-text', [
+            AlloyEvents.run<UpdateMenuTextEvent>(updateMenuText, (comp, se) => {
+              Representing.setValue(comp, se.event.text);
+            }),
+            AlloyEvents.run(NativeEvents.focusout(), (_comp, se) => {
+              spec.onAction(Value.get(se.event.target));
```

Comment:
I general, it's a bad idea to mix `Value.get` and `Representing.getValue`. If the problem is that they are out-of-sync, that you'd want to address that problem. But having `onEnter` get the value of the input in a different way to `focusout` and `change` ... that might cause problems.

Still with `Representing.getValue` wherever possible.

Reply:
changed